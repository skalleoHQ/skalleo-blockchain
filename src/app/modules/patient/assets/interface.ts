export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
}