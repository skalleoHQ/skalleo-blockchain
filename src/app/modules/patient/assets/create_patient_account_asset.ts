import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

const {
	VALID_PATIENT_DOMAIN,
	createPatientAccount,
	getAllPatientAccounts,
	setAllPatientAccounts,

} = require("../data/utils")



export class CreatePatientAccountAsset extends BaseAsset {
	public name = 'createPatientAccount';
  	public id = 1;

  // Define schema for asset
	public schema = {
    $id: 'patient/createPatientAccount-asset',
		title: 'CreatePatientAccountAsset transaction asset for patient module',
		type: 'object',
		required: ['patientIdentificationNumber', 'areaCode', 'username'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			username: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
  	};

	public validate({ asset }: ValidateAssetContext<{}>): void {
		// Verify if client fill patient Identification Number box
		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter your identification number');
		};

		// Implement checking identity procedure (redirect the client to national database for example and verify identity)

		/**
		 * */ 

		// Verify if client fill areaCode box
		if (!asset.areaCode) {
			throw new Error('You must enter your areaCode, you can find it on https://skalleo.io');
		};

		// Implement checking areaCode procedure (redirect the client to national database for example and verify if this zone is concerned)
		// OR verify it with apply we can look CHAIN_STATE_POOL_ACCOUNTS if the areaCode is available

		/**
		 * */
		

		// Verify if username is valid
		const chunks = asset.username.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid TLDs are "${VALID_PATIENT_DOMAIN.join()}"`);
		};

	

		

	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
		//create Patient account
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get(senderAddress);
		const areaCode = asset.areaCode;

		//Admit only one patient account per account
		if (senderAccount.patient.selfPatient) {
			throw new Error('You have already a patient account !')
		}

		//Verify if areaCode is valid
		if (!(areaCode in CHAIN_STATE_POOL_ACCOUNTS)) {
			throw new Error('Your area is not yet concerned, sorry !');
		}

		const patientAccount = createPatientAccount({
			patientIdentificationNumber: asset.patientIdentificationNumber,
        	areaCode: asset.areaCode,
        	username: asset.username,
        	ownerAddress: asset.ownerAddress,
			nonce: asset.nonce,
		});

		//update sender account with unique Patient ID
		senderAccount.patient.selfPatient = patientAccount;
		await stateStore.account.set(senderAddress, senderAccount);
		
		//save patient
		const allAccounts = await getAllPatientAccounts(stateStore);
		allAccounts.push(patientAccount);
		await setAllPatientAccounts(stateStore, allAccounts);






	}



}





