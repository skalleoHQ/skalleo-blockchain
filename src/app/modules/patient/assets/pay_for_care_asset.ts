import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

export class PayForCareAsset extends BaseAsset {
	public name = 'payForCare';
  public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'patient/payForCare-asset',
		title: 'PayForCareAsset transaction asset for patient module',
		type: 'object',
		required: [],
		properties: {},
  };

  public validate({ asset }: ValidateAssetContext<{}>): void {
    // Validate your asset
  }

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
		throw new Error('Asset "payForCare" apply hook is not implemented.');
	}
}
