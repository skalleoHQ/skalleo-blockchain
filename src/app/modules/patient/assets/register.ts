export interface PatientModuleProps {
    selfPatient: 'bytes';
    reverseLookup: 'bytes';
};


export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
};


export interface RegisteredPatientAccountsSchemaProps {
	registeredPatientAccounts: []
};



