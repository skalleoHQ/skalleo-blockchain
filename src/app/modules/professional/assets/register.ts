export interface ProfessionalModuleProps{
    selfProfessional: Buffer,
    reverseLookup: Buffer
};


export interface CreateProfessionalAccountAssetProps {
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
};


export interface RegisteredProfessionalAccountsSchemaProps {
    readonly registeredProfessionalAccounts: [];
};


export interface RegisteredProfessionalAccount {
    readonly id: Buffer;
    readonly ownerAddress: Buffer;
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
}




