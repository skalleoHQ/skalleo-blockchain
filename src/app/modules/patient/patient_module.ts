/* eslint-disable class-methods-use-this */

const { EMPTY_BUFFER } = require ('../patient/data/utils');

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, TransactionApplyContext
} from 'lisk-sdk';

import { CreatePatientAccountAsset } from "./assets/create_patient_account_asset";

export class PatientModule extends BaseModule {
    public accountSchema = {
        $id: 'skalleo/patient/patientAccount',
        type: 'object',
        required: ['selfPatient', 'reverseLookup'],
        properties: {
            reverseLookup: {
                dataType: 'bytes',
                fieldNumber: 1,
            },
            selfPatient: {
                dataType: 'bytes',
                fieldNumber: 2, 
            },
        },
        default: {
            selfPatient: EMPTY_BUFFER,
            reverseLookup: EMPTY_BUFFER,
        }
    }



    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'patient';
    public transactionAssets = [new CreatePatientAccountAsset()];
    public events = [
        // Example below
        // 'patient:newBlock',
    ];
    public id = 1000;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccounts = genesisBlock.header.asset.accounts;
    }
}
