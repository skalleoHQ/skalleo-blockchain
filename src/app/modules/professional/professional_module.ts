/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, TransactionApplyContext
} from 'lisk-sdk';

import { CreateProfessionalAccountAsset } from "./assets/create_professional_account_asset";

const { EMPTY_BUFFER, 

} = require ('../professional/data/utils');



export class ProfessionalModule extends BaseModule {
    public accountSchema = {
        $id: 'skalleo/professional/professionalAccount',
        type: 'object',
        required: ['selfProfessional', 'reverseLookup'],
        properties: {
            reverseLookup: {
                dataType: 'bytes',
                fieldNumber: 1,
            },
            selfProfessional: {
                dataType: 'bytes',
                fieldNumber: 2,
            },
        },
        default: {
            selfProfessional: EMPTY_BUFFER,
            reverseLookup: EMPTY_BUFFER,
        }
    };


    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'professional';
    public transactionAssets = [new CreateProfessionalAccountAsset()];
    public events = [
        // Example below
        // 'professional:newBlock',
    ];
    public id = 1001;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccounts = genesisBlock.header.asset.accounts;
    }
}
