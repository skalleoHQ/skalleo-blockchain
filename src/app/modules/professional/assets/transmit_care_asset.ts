import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { TransmitCareAssetProps } from './register';


const {
	VALID_PATIENT_DOMAIN,
	getAllPatientAccounts,

} = require ('../../patient/data/utils');



const {
	getAllProfessionalAccounts,

} = require ('./register');



const {
	recordCare,
	getAllRecordedCare,
	setAllRecordedCare,
	
} = require ('../data/care')



export class TransmitCareAsset extends BaseAsset<TransmitCareAssetProps> {
	public name = 'transmitCare';
  	public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'professional/transmitCare-asset',
		title: 'TransmitCareAsset transaction asset for professional module',
		type: 'object',
		required: ['patientIdentificationNumber', 'reverseLookup', 'areaCode', 'careSpecifications'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			reverseLookup: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3
			},
			careSpecifications: {
				dataType: 'string',
				fieldNumber: 4,
			},
		},
	};

	public validate({ asset }: ValidateAssetContext<TransmitCareAssetProps>): void {

		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter identification number of your patient');
		}

		if (!asset.reverseLookup) {
			throw new Error('You must enter username of your patient');
		}

		const chunks = asset.reverseLookup.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PATIENT_DOMAIN.join()}"`);
		};

		if (!asset.areaCode) {
			throw new Error('You must enter areaCode of your patient');
		}

		//This careSpecifications will be used to update patient's medical record in his national database and plugin database
		if (!asset.careSpecifications) {
			throw new Error('Please update the cares specifications of your patient !')
		}

		//Must implement procedure in order to send specifications to national database & verify if careSpecs respect skeleton

	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<TransmitCareAssetProps>): Promise<void> {
		const senderAddress = transaction.senderAddress;
		
		//Verify if patient exists
		const patientAccounts = await getAllPatientAccounts(stateStore);

		const patientIdentificationNumberIndex = patientAccounts.findIndex((t) => t.patientIdentificationNumber === asset.patientIdentificationNumber);
		if (patientIdentificationNumberIndex < 0) {
			throw new Error('Patient not found !');	
		}

		const patientReverseLookupIndex = patientAccounts.findIndex((t) => t.username === asset.reverseLookup);
		if (patientReverseLookupIndex < 0) {
			throw new Error('Patient not found !');
		}

		/*const patientAreaCodeIndex = patientAccounts.findIndex((t) => t.areaCode === asset.areaCode);
		if (patientAreaCodeIndex < 0) {
			throw new Error('Patient not found !');
		}*/
		//assert(patientReverseLookupIndex === patientAreaCodeIndex); impossible because areaCode not unique

		if(patientIdentificationNumberIndex !== patientReverseLookupIndex) {
			throw new Error('Patient not found !');
		}

		const professionalAccounts = await getAllProfessionalAccounts(stateStore);
		const addressIndex = professionalAccounts.findIndex((t) => t.ownerAddress === transaction.senderAddress);
		
		const professionalAccount = professionalAccounts[addressIndex];
		const patientAccount = patientAccounts[patientIdentificationNumberIndex]; 

		if (professionalAccount.ownerAddress === patientAccount.ownerAddress) {
			throw new Error('Transaction error !');
		}
		
		//create care for a patient
		const care = recordCare({
			patientIdentificationNumberIndex: asset.patientIdentificationNumber,
			reverseLookup: asset.reverseLookup,
			areaCode: asset.areaCode,
			careSpecifications: asset.careSpecifications,
			senderAddress: senderAddress,
			nonce: transaction.nonce,
		})

		//update care data chain
		const allCare = await getAllRecordedCare(stateStore);
		allCare.push(care);
		await setAllRecordedCare(stateStore, allCare);

	}


}



module.exports = {
	TransmitCareAsset,
}
