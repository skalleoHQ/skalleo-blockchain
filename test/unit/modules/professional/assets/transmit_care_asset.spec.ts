import { StateStore, testing } from 'lisk-sdk';
import { CreatePatientAccountAsset } from '../../../../../src/app/modules/patient/assets/create_patient_account_asset';
import { PatientModuleProps } from '../../../../../src/app/modules/patient/assets/register';
import { PatientModule } from '../../../../../src/app/modules/patient/patient_module';
import { ProfessionalModuleProps } from '../../../../../src/app/modules/professional/assets/register';
import { TransmitCareAsset } from '../../../../../src/app/modules/professional/assets/transmit_care_asset';
import { ProfessionalModule } from '../../../../../src/app/modules/professional/professional_module';

describe('TransmitCareAsset', () => {
  let transactionAsset: TransmitCareAsset;
  let createPatientAccountAsset: CreatePatientAccountAsset;

	beforeEach(() => {
		transactionAsset = new TransmitCareAsset();
		createPatientAccountAsset = new CreatePatientAccountAsset(); 
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(2);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('transmitCare');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it('should throw error if patientIdentificationNumber box is empty', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "", reverseLookup: "moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter identification number of your patient'
				)
			});

			it('should throw error if reverseLookup is empty', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter username of your patient'
				)
			});

			it('should throw error if areaCode is empty', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "moussa.adh", areaCode: "", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You must enter areaCode of your patient'
				)
			});

			it('should throw error if careSpecs is empty', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'Please update the cares specifications of your patient !'
				)
			});

			it('should throw error if reverseLookup is not right', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "by.moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'You may not use "." than for the domain'
				)
			});

			it('should throw error if reverseLookup is not right', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "by.moussa", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).toThrow(
					'Invalid domain found "moussa". Valid domain is "adh"'
				)
			});

			it('should be ok for valid schema', () => {
				const context = testing.createValidateAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				expect(() => transactionAsset.validate(context)).not.toThrow();
			});

		});
	});

	describe('apply', () => {
		let stateStore: StateStore;
		let account: any;
		let account1: any;
		let proAccount: any;
		let proAccount1: any;

		beforeEach(() => {
			account = testing.fixtures.createDefaultAccount<PatientModuleProps>([PatientModule]);
			account1 = testing.fixtures.createDefaultAccount<PatientModuleProps>([PatientModule]);
			proAccount = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);
			proAccount1 = testing.fixtures.createDefaultAccount<ProfessionalModuleProps>([ProfessionalModule]);

			stateStore = new testing.mocks.StateStoreMock({
				accounts: [account, account1, proAccount, proAccount1],
			});

			jest.spyOn(stateStore.chain, 'get');
			jest.spyOn(stateStore.chain, 'set');
		})

		describe('valid cases', () => {
			it('should update the state store with the care transmitted', async() => {
				const context = testing.createApplyAssetContext({
					stateStore,
					asset: { patientIdentificationNumber: "3301012022", areaCode: "221_SENEGAL", username: "moussa.adh" },
					transaction: { senderAddress: account.address } as any,
				});

				await createPatientAccountAsset.apply(context);

				const context1 = testing.createApplyAssetContext({
					stateStore,
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: proAccount.address } as any
				})

				await transactionAsset.apply(context1);
				
				expect(stateStore.chain.networkIdentifier).toEqual(context1.stateStore.chain.networkIdentifier);
				expect(stateStore.chain.set).toBe(context1.stateStore.chain.set);
				expect(stateStore.chain.set).toHaveBeenCalledWith("professional: recordedCare", expect.any(Buffer))

			});
		});

		describe('invalid cases', () => {
			/*it('should throw error patient not found', async() => {
				const context = testing.createApplyAssetContext({
					asset: { patientIdentificationNumber: "3301012022", reverseLookup: "moussa.adh", areaCode: "221_SENEGAL", 
						careSpecifications: "Juste un examen de prévention, Conclusion: le patient ne présente aucun problème" },
					transaction: { senderAddress: Buffer.alloc(0) } as any
				})

				await transactionAsset.apply(context);

				await expect(() => transactionAsset.apply(context)).rejects.toThrow(
					'Patient not found !'
				)
			});*/



		});

	});

});
