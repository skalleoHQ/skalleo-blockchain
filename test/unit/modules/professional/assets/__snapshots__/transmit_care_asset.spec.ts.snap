// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransmitCareAsset constructor should have valid schema 1`] = `
Object {
  "$id": "professional/transmitCare-asset",
  "properties": Object {
    "areaCode": Object {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "careSpecifications": Object {
      "dataType": "string",
      "fieldNumber": 4,
    },
    "patientIdentificationNumber": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
    "reverseLookup": Object {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "required": Array [
    "patientIdentificationNumber",
    "reverseLookup",
    "areaCode",
    "careSpecifications",
  ],
  "title": "TransmitCareAsset transaction asset for professional module",
  "type": "object",
}
`;
